src/prefect_server/utilities/collections.py

```diff
@@ -19,3 +20,61 @@ def chunked_iterable(iterable: Iterable, size: int):
         if not chunk:
             break
         yield chunk
+
+
+class TimedUniqueValueStore:
+    """
+    Stores unique values for a specific amount of time.
+
+    Args:
+        - duration (float, int): the number of seconds to store a value before removing it
+    """
+
+    def __init__(self, duration: Union[float, int] = 10):
+        self.values = set()
+        self.lock = threading.Lock()
+        self.duration = duration
+
+    def add_and_expire(self, value: Hashable, duration: Optional[float, int] = None) -> bool:
+        """
+        Add a unique value to store and kick off a timer to remove it
+
+        Args:
+            - value (Hashable): value to store, must be hashable and
+                able to be stored in a set
+            - duration (float, int, optional): the number of seconds to store a value
+                before removing it. Defaults to self.duration if not provided
+        Returns:
+            - bool: `True` if the value did not already exist in the store
+                and is successfully added, otherwise `False`
+        """
+        if duration is None:
+            duration = self.duration
+
+        with self.lock:
```

Comment:
Are there any edge cases where we may want to put on a timeout on the lock acquisition?  E.g., instead of using the context manager do something like:
```python
acquired = self.lock.acquire(timeout=2)
try:
    if not acquired:
        return False
    ... # value check, etc.
finally:
    self.lock.release()

return bool_val
```

the downside ofc is that it's more complicated

Reply:
We could write a custom context manager that handles this and use it instead of the stdlib one (if we think it's a good idea)

Reply:
Unless we have a concrete case, I don't think we need to do this. I've seen lots of usage of the stdlib context manager without timeouts.

Reply:
I'll leave it without for now, but can always go back and add